x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationNodeService.java

```diff
@@ -0,0 +1,387 @@
+                r -> logger.debug(() -> new ParameterizedMessage("[{}] model loaded and accepting routes", modelId)),
+                e -> {
+                    // This means that either the allocation has been deleted, or this node's particular route has been removed
+                    if (ExceptionsHelper.unwrapCause(e) instanceof ResourceNotFoundException) {
+                        logger.debug(
+                            () -> new ParameterizedMessage(
+                                "[{}] model loaded but failed to start accepting routes as allocation to this node was removed",
+                                modelId
+                            ),
+                            e
+                        );
+                    }
+                    // this is an unexpected error
+                    logger.warn(() -> new ParameterizedMessage("[{}] model loaded but failed to start accepting routes", modelId), e);
+                }
+            )
+        );
+    }
+
+    private void updateStoredState(
+        String modelId,
+        RoutingStateAndReason routingStateAndReason,
+        ActionListener<AcknowledgedResponse> listener
+    ) {
+        if (stopped) {
+            return;
+        }
+        trainedModelAllocationService.updateModelAllocationState(
+            new UpdateTrainedModelAllocationStateAction.Request(nodeId, modelId, routingStateAndReason),
+            ActionListener.wrap(success -> {
+                logger.debug(
+                    () -> new ParameterizedMessage("[{}] model is [{}] and master notified", modelId, routingStateAndReason.getState())
+                );
+                listener.onResponse(AcknowledgedResponse.TRUE);
+            },
+                // TODO: error could be that the model is not allocated to this node any longer or that the master is
+                // TODO: Re-do on master notification failure if its due to master not elected yet
+                error -> {
+                    logger.warn(
+                        () -> new ParameterizedMessage(
+                            "[{}] model is [{}] but failed to notify master",
+                            modelId,
+                            routingStateAndReason.getState()
+                        ),
+                        error
+                    );
+                    listener.onFailure(error);
+                }
+            )
+        );
+    }
+
+    private void handleLoadFailure(TrainedModelDeploymentTask task, Exception ex) {
+        logger.error(() -> new ParameterizedMessage("[{}] model failed to load", task.getModelId()), ex);
+        if (task.isStopped()) {
+            logger.debug(() -> new ParameterizedMessage("[{}] model failed to load, but is now stopped", task.getModelId()));
+        }
+        // TODO: Do we want to remove from the modelIdToTask map? This would cause it to be reloaded by state updates on INITIALIZING
+        modelIdToTask.remove(task.getModelId());
```

Comment:
If the node state in the cluster state somehow got set back to initializing (of which there is no way now other than recreating the allocation or deleting the route), we would start the model again, which may be exactly what we want.